src/twisted/web/test/test_wsgi.py

```diff
@@ -1941,6 +1943,68 @@ def cbRendered(ignored):
         return d
 
 
+    @inlineCallbacks
+    def test_pendingRequestOperationsRespectConnectionLoss(self):
+        """
+        If there are pending transport writes after
+        C{Request.notifyFinish} is called, the transport should
+        not be written to.
+        """
+
+        self.enableThreads()
+        trackedRequest = [None]
+
+        def applicationFactory():
+            def application(environ, startResponse):
+                startResponse('200 OK', [])
+                self.reactor.callFromThread(
+                    trackedRequest[0].connectionLost,
+                    Failure(ConnectionLost("No more connection")))
+                yield b'write'
+
+            return application
+
+        def requestFactory(*args, **kwargs):
+            # We need to manually close the request instantiated
+            # by self.lowLevelRender().
+            self.assertIsNone(trackedRequest[0])
+            trackedRequest[0] = Request(*args, **kwargs)
+            return trackedRequest[0]
+
+        # Keep a copy of _WSGIResponse.write before we patch it.
+        _write = _WSGIResponse.write
+
+        # Signal for the delayed write allowing it to proceed.
+        connectionLost = threading.Event()
+
+        # Result of the (delayed) write.
+        writeResult = Deferred()
+
+        def _WSGIResponse_write(self, data):
+            # Block write until transport is closed.
+            connectionLost.wait()
+
+            try:
+                r = _write(self, data)
+                self.reactor.callFromThread(writeResult.callback, None)
+                return r
+            except:
+                self.reactor.callFromThread(writeResult.errback, Failure())
+
+        self.patch(_WSGIResponse, 'write', _WSGIResponse_write)
+
+        # Run and wait for the request to fail.
+        request = self.lowLevelRender(
+            requestFactory, applicationFactory, DummyChannel,
+            'GET', '1.1', [], [''])
+
+        yield self.failUnlessFailure(request.notifyFinish(), ConnectionLost)
+
+        # Unblock write and wait for it to finish.
+        connectionLost.set()
+        yield writeResult
```

Comment:
I think that we should also check what was written to the network, not only that write was successful.

it might be successful,  but have some extra info written to the network

Reply:
The point is what it wasn't written, because the write was only attempted after the connection was lost. The original bug was that the attempted write did not check whether the transport was still open (and subsequently failed).

Reply:
but maybe test that transport is empty... or update the implementation to yield some data before the sync point

in this way we can check that only the initial data was written and no other unwanted data.

Reply:
I check now that the write was attempted, and that nothing was actually written to the transport.